### PR TITLE
Added discrete wavelet transform processor 

### PR DESCRIPTION
**Pull request by esleon97**
Before submitting a pull request, please make sure you've read and understood the pygama developer's guide: https://pygama.readthedocs.io/en/latest/developer.html. In particular, do not forget to:

- [ ] Conform to our **coding conventions**
- [ ] Update existing or add new **tests**
- [ ] Update existing or add new **documentation**
- [ ] Address any issue reported by GitHub checks

### CODE REVIEW

File: src/pygama/dsp/processors/dwt.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def dwt(wave_type: str, level: int) -> Callable:
+    """
+    Apply a discrete wavelet transform to the waveform and return only
+    the approximate coefficients.
+
+    Note
+    ----
+    This processor is composed of a factory function that is called using the
+    'init_args' argument. The input and output waveforms are passed using
+    'args'. The output waveform dimension must be specified.
```
Comment (gipert) [inline src/pygama/dsp/processors/dwt.py:21]:
```suggestion
    This processor is composed of a factory function that is called using the
    ``init_args`` argument. The input and output waveforms are passed using
    ``args``. The output waveform dimension must be specified.
```

File: src/pygama/dsp/processors/dwt.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+    Parameters
+    ----------
+    wave_type
+       The wavelet type for discrete convolution ('haar', 'db1', ...)
+    level
+       The level of decompositions to be performed (1, 2, ...)
```
Comment (gipert) [inline src/pygama/dsp/processors/dwt.py:29]:
```suggestion
       The wavelet type for discrete convolution ``('haar', 'db1', ...)``
    level
       The level of decompositions to be performed ``(1, 2, ...)``
```

File: src/pygama/dsp/processors/dwt.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    def dwt_out(wf_in: np.ndarray, wf_out: np.ndarray) -> None:
+        """
+        Parameters
+        ----------
+        wf_in
+           The input waveform
+        wf_out
+           The approximate coefficients. The dimension of this array can be calculated
+           by out_dim = wf_length/(filter_length^level), where filter_length can be
+           obtained via pywt.Wavelet(wave_type).dec_len
```
Comment (gipert) [inline src/pygama/dsp/processors/dwt.py:63]:
```suggestion
           The approximate coefficients. The dimension of this array can be calculated
           by ``out_dim = wf_length/(filter_length^level)``, where ``filter_length`` can be
           obtained via ``pywt.Wavelet(wave_type).dec_len``.
```

**Review by gipert — CHANGES_REQUESTED** [review]
Thanks a lot! I just left some suggestions about doctring styling.

**Review by gipert — APPROVED** [review]

**Review by ggmarshall — COMMENTED** [review]
Looks good, to conform with the other processors can you change it from wf_in/wf_out to w_in/w_out, do you want to add it to the init file also

**Comment by gipert** [comment]
@esleon97 can you please update this according to https://github.com/legend-exp/pygama/pull/350?

**Review by ggmarshall — APPROVED** [review]

**Comment by ggmarshall** [comment]
It seems the pre-commit removed the changes to the init file?

**Comment by gipert** [comment]
You need to add the processor name to the `__all__` vector in `__init__.py`.

Another thing: `dwt()` is maybe a bit cryptic, why not `discrete_wavelet_transform()`? Or just `discrete_wavelet()`?

**Comment by jasondet** [comment]
@gipert looks like everything is dying on this now. Was whatever is responsible for flake8 somehow updated under the hood? Anyway, I'll see if I can fix it...

**Comment by gipert** [comment]
Not sure but looks like a new flake8 feature. Guess we can just add those decorators...

**Comment by jasondet** [comment]
flake8 error is fixed in https://github.com/legend-exp/pygama/pull/373, will pull it in then merge this

**Comment by gipert** [comment]
@jasondet can you also squash the commits before merging? I think we should do it always (if it makes sense) to keep our Git history more linear and clean.

**Pull request merged**